scipy/stats/stats.py

```diff
@@ -4153,6 +4154,178 @@ def binary_search(n, n1, n2, side):
+        Number of iterations of the grid search. Default is 4. Must be
+        non-negative.
+
+    Returns
+    -------
+    score_statistic : float
+        Z statistic with pooled (Score) or unpooled (Wald) variance, depending
+        on the user choice of `pooled` param.
+    p_value : float
+        P-value, the probability of obtaining a distribution at least as
+        extreme as the one that was actually observed, assuming that the
+        null hypothesis is true.
+
+    See Also
+    --------
+    chi2_contingency : Chi-square test of independence of variables in a
+        contingency table.
+    fisher_exact : Fisher exact test on a 2x2 contingency table.
+
+    References
+    ----------
+    .. [1] `Mehta, Cyrus & Senchaudhuri, Pralay. (2003). Conditional
+    versus Unconditional Exact Tests for Comparing Two Binomials`__.
+
+    .. __: https://www.statsols.com/hubfs/Resources_/Comparing-Two-Binomials.pdf
+
+    Examples
+    --------
+    examples from [1]_.
+
+    Consider the following example of a vaccine efficacy study (Chan, 1998).
+    In a randomized clinical trial of 30 subjects, 15 were innoculated with a
+    recombinant DNA influenza vaccine and the 15 were innoculated with a
+    placebo. Twelve of the 15 subjects in the placebo group (80%) eventually
+    became infected with influenza whereas for the vaccine group, only 7 of
+    the 15 subjects (47%) became infected. The data are tabulated as a 2 × 2
+    table::
+
+                    Vaccine  Placebo
+        Yes             8        2
+        No              1        5
+
+    We use this table to find the p-value:
+    >>> import scipy.stats as stats
+    >>> TX, pvalue = stats.barnard_exact([[7, 12], [8, 3]])
+    >>> pvalue
+    0.0681...
+
+    """
+    if num_it < 0:
+        raise ValueError(
+            "Number of iterations must be non-negative, "
+            f"found {num_it!r}"
+        )
+
+    c = np.asarray(
+        table,
+        dtype=np.int64)
+
+    if not c.shape == (2, 2):
+        raise ValueError("The input `table` must be of shape (2, 2).")
+
+    if np.any(c < 0):
+        raise ValueError("All values in `table` must be nonnegative.")
+
+    if 0 in c.sum(axis=0):
+        # If both values in a row or column are zero, the p-value is 1 and
+        # the score's statistic is NaN.
+        return np.nan, 1.0
+
+    total_c1, total_c2 = c.sum(axis=0)
+    n = total_c1 + total_c2
+
+    x1 = np.arange(total_c1 + 1, dtype=np.int64).reshape(-1, 1)
+    x2 = np.arange(total_c2 + 1, dtype=np.int64).reshape(1, -1)
+
+    p1, p2 = x1 / total_c1, x2 / total_c2
+
+    if pooled:
+        p = (x1 + x2) / (total_c1 + total_c2)
+        var_p1_p2 = p * (1 - p) * (1 / total_c1 + 1 / total_c2)
+    else:
+        var_p1_p2 = p1 * (1 - p1) / total_c1 + p2 * (1 - p2) / total_c2
+
+    # To avoid warning when dividing by 0
+    with np.errstate(divide='ignore', invalid='ignore'):
+        TX = np.divide((p2 - p1), np.sqrt(var_p1_p2))
```

Comment:
You can add after this block a line like:

```python
TX[var_p1_p2==0] = (p2-p1)[var_p1_p2==0] * np.inf
```

With this, you have ±∞, allowing to keep indexes in pvalue computation. Also, this should solve the problem where one test of the CI fails.

Reply:
Done

Reply:
That appear to have been incorrect:

```
TX[var_p1_p2 == 0] = (p2-p1) * [var_p1_p2 == 0] * np.inf
E   TypeError: NumPy boolean array indexing assignment requires a 0 or 1-dimensional input, input has 3 dimensions
        TX         = array([[        nan,  1.01709526,  1.46385011,  1.82574186,  2.14834462,
         2.44948974,  2.73861279,  3.02166093....02166093, -2.73861279,
        -2.44948974, -2.14834462, -1.82574186, -1.46385011, -1.01709526,
                nan]])
        alternative = 'two-sided'
        c          = array([[ 7, 12],
       [ 8,  3]])
        n          = 30
        num_it     = 4
        p          = array([[0.        , 0.03333333, 0.06666667, 0.1       , 0.13333333,
        0.16666667, 0.2       , 0.23333333, 0.2666...3333, 0.76666667, 0.8       ,
        0.83333333, 0.86666667, 0.9       , 0.93333333, 0.96666667,
        1.        ]])
        p1         = array([[0.        ],
       [0.06666667],
       [0.13333333],
       [0.2       ],
       [0.26666667],
       [0.333...[0.66666667],
       [0.73333333],
       [0.8       ],
       [0.86666667],
       [0.93333333],
       [1.        ]])
        p2         = array([[0.        , 0.06666667, 0.13333333, 0.2       , 0.26666667,
        0.33333333, 0.4       , 0.46666667, 0.53333333, 0.6       ,
        0.66666667, 0.73333333, 0.8       , 0.86666667, 0.93333333,
        1.        ]])
        pooled     = True
        table      = [[7, 12], [8, 3]]
        total_c1   = 15
        total_c2   = 15
        var_p1_p2  = array([[0.        , 0.0042963 , 0.0082963 , 0.012     , 0.01540741,
        0.01851852, 0.02133333, 0.02385185, 0.0260...7407, 0.02385185, 0.02133333,
        0.01851852, 0.01540741, 0.012     , 0.0082963 , 0.0042963 ,
        0.        ]])
        x1         = array([[ 0],
       [ 1],
       [ 2],
       [ 3],
       [ 4],
       [ 5],
       [ 6],
       [ 7],
       [ 8],
       [ 9],
       [10],
       [11],
       [12],
       [13],
       [14],
       [15]])
        x2         = array([[ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15]])
```